### PR TITLE
Support using Git binary for some operations

### PR DESCRIPTION
**Pull request by adityasaky**
Related: #2

### CODE REVIEW

**Comment by adityasaky** [comment]
@wlynch this PR has a large diff but it's largely in tests. I've reimplemented three gitinterface APIs using the Git binary in place of go-git. They are GetFilePathsChangedByCommitUsingBinary, KnowsCommitUsingBinary, and GetCommitsBetweenRangeUsingBinary. Note that most of the diff is because I've copied the tests for the go-git equivalents. The reason the tests aren't the same is because with the Git binary, we can't use an in-memory repo. We could, if you prefer, switch the go-git tests to use on-disk repo as the test cases for all three are otherwise identical (caveat: some errors returned differ).

**Comment by adityasaky** [comment]
With this feature flag enabled, we see some significant performance improvements when verifying file policies. It'll be interesting to see if we can help go-git with improvements by comparing the two. On https://github.com/gittuf/benchmark-file-rules, we see the following when using go-git:

```
$ time gittuf verify-ref --from-entry b3f629d7d1fcd951449765ebc3336c982dcecde7 main

real    0m7.431s
user    0m8.800s
sys     0m0.875s
```

When we switch to the Git binary:

```
$ time gittuf --use-git-binary verify-ref --from-entry b3f629d7d1fcd951449765ebc3336c982dcecde7 main

real    0m0.676s
user    0m0.843s
sys     0m0.072s
```

File: internal/rsl/rsl.go [hunk 1]
```diff
@@ -328,7 +327,20 @@ func GetLatestEntry(repo *git.Repository) (Entry, error) {
...
+	var (
+		it  Entry
+		err error
+	)
+
+	if anchor.IsZero() {
+		it, err = GetLatestEntry(repo)
+	} else {
+		it, err = GetEntry(repo, anchor)
+	}
```
Comment (wlynch) [inline internal/rsl/rsl.go:343]: This looks like an unrelated diff. Is this rebased on another change?

**Review by wlynch — COMMENTED** [review]
I'd be curious to see some of the profiles for the large repo case. I suspect that we're probably holding go-git naively that might explain some of the slowdown 🤔 

If both implementations were reading the same files, I wouldn't expect a slow down of this order of magnitude.

File: internal/gitinterface/changes.go [hunk 2]
```diff
@@ -40,7 +45,16 @@ func GetCommitFilePaths(commit *object.Commit) ([]string, error) {
 // one parent, check if the commit is the same as at least one of its parents.
 // If there is a matching parent, we return no changes. If there is no matching
 // parent commit, we return the changes between the commit and each of its parents.
-func GetFilePathsChangedByCommit(repo *git.Repository, commit *object.Commit) ([]string, error) {
+func GetFilePathsChangedByCommit(repo *git.Repository, commitID plumbing.Hash) ([]string, error) {
+	if featureflags.UseGitBinary {
+		return GetFilePathsChangedByCommitUsingBinary(repo, commitID)
+	}
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/gitinterface/changes.go:51]: This is what I originally wanted to do but I wasn't sure if the diff / package renames it'd cause across packages was worth it. I'm happy to take a stab at this, in which case we should turn this into a draft and leave it as is for the time being.

File: internal/cmd/root/root.go [hunk 3]
```diff
@@ -55,6 +59,13 @@ func (o *options) AddFlags(cmd *cobra.Command) {
 		"memory.prof",
 		"file to store memory profile",
 	)
+
+	cmd.PersistentFlags().BoolVar(
+		&o.useGitBinary,
+		"use-git-binary",
+		false,
+		fmt.Sprintf("use Git binary for some operations (developer mode only, set %s=1)", devmode.DevModeKey),
+	)
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/cmd/root/root.go:68]: I think we ought to discuss. I think having a go-git instance makes it easier to test some things, but if we want users to primarily use the git binary backend long term, we'll have to test that too. Maybe we ought to take a step back and discuss the roadmap here. cc @JustinCappos

File: internal/rsl/rsl.go (hunk 1 above)
Comment (adityasaky) [reply internal/rsl/rsl.go:343]: rip, yes. sorry, will fix if we decide to proceed with this PR!

File: internal/cmd/root/root.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adityasaky) [reply internal/cmd/root/root.go:68]: Note that switching over to the binary would simplify some features like command compatibility with Git, I think.

**Comment by pjbgf** [comment]
@adityasaky it would be great to get a pprof profile for this specific process, so that we can identify bottlenecks in go-git.

**Comment by adityasaky** [comment]
@pjbgf apologies for the delay, I'll try and capture those ASAP. We added support for profiling recently (#284).

**Comment by adityasaky** [comment]
Closing in favour of #367
